hypershift-operator/controllers/hostedcluster/internal/platform/core.go

```diff
@@ -0,0 +1,60 @@
+package platform
+
+import (
+	"context"
+	"fmt"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/upsert"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	hostedClusterAnnotation = "hypershift.openshift.io/cluster"
+)
+
+type Platform interface {
+	// ReconcileCAPIInfraCR is responsible for creating or updating a CAPIInfraCR.
+	ReconcileCAPIInfraCR(hcluster *hyperv1.HostedCluster, name, controlPlaneNamespace string, apiEndpoint hyperv1.APIEndpoint,
+		c client.Client, ctx context.Context) (client.Object, error)
+
+	// CAPIProviderDeploymentSpec should return a CAPIProvider DeploymentSpec.
+	CAPIProviderDeploymentSpec(controlPlaneNamespace, tokenMinterImage string) (*appsv1.DeploymentSpec, error)
+
+	// HCPPlatformSpec should return a hyperv1.PlatformSpec for the HCP.
+	HCPPlatformSpec(hcp *hyperv1.PlatformSpec, controlPlaneNamespace string) error
+
+	// ReconcileCredentials is responsible for creating or updating resources related to cloud credentials.
+	ReconcileCredentials(c client.Client, ctx context.Context, createOrUpdate upsert.CreateOrUpdateFN,
```

Comment:
As with https://github.com/openshift/hypershift/pull/719/files#r758636418, what if the contract were for the user to produce a specific set of outputs to be persisted on the controller side? Otherwise it's not clear how this is not simply arbitrary behavior injected into the controller at a certain point in the lifecycle.

Reply:
>Otherwise it's not clear how this is not simply arbitrary behavior injected into the controller at a certain point in the lifecycle.

For ReconcileCredentials it is actually kind of that. This manipulate platform specific API fields that might differ across platforms and therefore I think is too soon to have a clear input/output contract.